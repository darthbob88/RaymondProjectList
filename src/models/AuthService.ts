import { firebaseAuth } from "../firebase"

export const doCreateUserWithEmailAndPassword = (email: string, password: string, userName: string) =>
    firebaseAuth.createUserWithEmailAndPassword(email, password).then(result => {
        if (result.user == null) return false
        result.user.updateProfile({
            displayName: userName
        }).catch(function (error) {
            console.log(error);
        })
    });

export const doSignInWithEmailAndPassword = (email: string, password: string) =>
    firebaseAuth.signInWithEmailAndPassword(email, password);

export const doSignOut = () => firebaseAuth.signOut();

export const doPasswordReset = (email: string) => firebaseAuth.sendPasswordResetEmail(email);

export const doPasswordUpdate = (password: string) => {
    if (firebaseAuth.currentUser != null) {
        firebaseAuth.currentUser.updatePassword(password);
    } else {
        //TODO: If not logged in, throw an error?
    }
}

export const setDisplayName = (userName: string) => {
    if (firebaseAuth.currentUser != null) {
        firebaseAuth.currentUser.updateProfile({ displayName: userName }).then(() => console.log("Successfully set displayName"))

    }
}