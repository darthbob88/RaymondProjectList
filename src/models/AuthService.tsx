import { firebaseAuth } from "../firebase";
import React, { useEffect, useState } from "react";

const doCreateUserWithEmailAndPassword = (
  email: string,
  password: string,
  userName: string
) =>
  firebaseAuth.createUserWithEmailAndPassword(email, password).then(result => {
    if (result.user == null) return false;
    result.user
      .updateProfile({
        displayName: userName
      })
      .catch(function(error) {
        console.log(error);
      });
  });

const doSignInWithEmailAndPassword = (email: string, password: string) =>
  firebaseAuth.signInWithEmailAndPassword(email, password);

const doSignOut = () => firebaseAuth.signOut();

const doPasswordReset = (email: string) =>
  firebaseAuth.sendPasswordResetEmail(email);

const doPasswordUpdate = (password: string) => {
  if (firebaseAuth.currentUser != null) {
    firebaseAuth.currentUser.updatePassword(password);
  } else {
    //TODO: If not logged in, throw an error?
  }
};

const currentUser = null;
export const AuthService = {
  currentUser,
  doPasswordReset,
  doCreateUserWithEmailAndPassword,
  doPasswordUpdate,
  doSignInWithEmailAndPassword,
  doSignOut
};

export const AuthContext = React.createContext(AuthService);

type AuthProviderTypes = { children: React.ReactNode };
export const AuthProvider: React.FunctionComponent<AuthProviderTypes> = ({
  children
}) => {
  const [currentUser, setCurrentUser] = useState();
  useEffect(() => {
    firebaseAuth.onAuthStateChanged(user => {
      setCurrentUser(user);
      console.log(`AuthStateChanged, current user is ${currentUser}`);
    });
  });
  return (
    <AuthContext.Provider value={{ ...AuthService, currentUser }}>
      {children}
    </AuthContext.Provider>
  );
};
