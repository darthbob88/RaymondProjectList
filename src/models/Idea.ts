export default interface Idea {
    summary: string;
    description: string;
    slug: string;
}

export const defaultIdeaList: Idea[] = [
    { description: "butts", summary: "butts", slug: "butts" },
{ description: "foo", summary: "bar", slug: "baz" }];