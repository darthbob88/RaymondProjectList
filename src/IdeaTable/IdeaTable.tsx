import React from "react";
import Idea from "../models/Idea";
import { Link, NavLink } from "react-router-dom";

type IdeaTableProps = {
  listOfIdeas: Idea[];
};
/**
 * Simple table to list ideas for me to work on. Will eventually fix it to show only summaries
 * until somebody clicks an idea to expand it, but for now this will do.
 * @param {Idea[]} listOfIdeas - An array of Idea to show
 * @component
 * @example
 * const ideaList = defaultIdeaList
 * return (
 *   <IdeaTable listOfIdeas={ideaList} />
 * )
 */
const IdeaTable: React.FunctionComponent<IdeaTableProps> = ({
  listOfIdeas
}) => {
  return (
    <React.Fragment>
      <h1>List of Ideas</h1>
      <ul id="ideaTable">
        {listOfIdeas.map(idea => (
     <Link to={`/ideas/${idea.slug}`}> <IdeaListing key={idea.slug} idea={idea} /></Link>
        ))}
      </ul>
    </React.Fragment>
  );
};

type IdeaListingProps = { idea: Idea };
/**TODO: Better name. */
const IdeaListing: React.FunctionComponent<IdeaListingProps> = ({ idea }) => (
  <li className="ideaListing">
    <h3>Summary: {idea.summary} </h3>
    <p>
      <b>Description</b>: {idea.description}
    </p>
    <p>
      <b>Status</b>: {idea.currentStatus}
    </p>
  </li>
);

export default IdeaTable;
