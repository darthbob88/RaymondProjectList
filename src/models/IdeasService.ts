import { databaseRef } from "../firebase";
import Idea from "./Idea";

//TODO: Turn this into a real-time watcher, so it gets automatically updated.
export const fetchAllIdeas = async (): Promise<Idea[]> => {
  let listOfIdeas: Idea[] = [];
  const querySnapshot = await databaseRef.collection("ideas").get();
  listOfIdeas = querySnapshot.docs.map(doc => {
    const newIdea = { ...doc.data(), id: doc.id } as Idea;
    return newIdea;
  });
  return listOfIdeas;
};
//TODO: Should this return a single idea, or the entire updated list?
export const addIdea = async (newIdea: Idea): Promise<Idea> => {
  await databaseRef
    .collection("ideas")
    .doc(newIdea.slug)
    .set(newIdea, { merge: true });
  const addedIdea = await databaseRef
    .collection("ideas")
    .doc(newIdea.slug)
    .get();
  return addedIdea.data() as Idea;
};
export const fetchSpecificIdea = async (ideaID: string): Promise<Idea> => {
  //TODO: Maybe I should abstract this to just ideasDB so I don't have to keep writing it out.
  const docRef = databaseRef.collection("ideas").doc(ideaID);
  let ideaSnapshot = await docRef.get();

  if (ideaSnapshot.exists) {
    // console.log("Document data:", ideaSnapshot.data());
    return ideaSnapshot.data() as Idea;
  } else {
    // doc.data() will be undefined in this case
    // console.log("No such document!");
    throw new Error("No such document!");
  }
};
