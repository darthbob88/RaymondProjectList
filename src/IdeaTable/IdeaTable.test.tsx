import React from 'react';
import ReactDOM from 'react-dom';
import IdeaTable from './IdeaTable';
import { defaultIdeaList } from '../models/Idea';
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from 'react-router-dom';


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Router> <IdeaTable listOfIdeas={defaultIdeaList} /></Router>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer
    .create(<Router> <IdeaTable listOfIdeas={defaultIdeaList} /></Router>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});