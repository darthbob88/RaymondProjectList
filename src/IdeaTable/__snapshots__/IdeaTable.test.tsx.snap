// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  " ",
  <h1>
    List of Ideas
  </h1>,
  <ul
    id="ideaTable"
  >
    <li
      className="ideaListing"
    >
      <h3>
        Summary: 
        Set up portfolio
         
      </h3>
      <p>
        <b>
          Description
        </b>
        : 
        Set up personal website, mostly for this but also for hosting Calibre
      </p>
      <p>
        <b>
          Status
        </b>
        : 
        To Do
      </p>
      <a
        href="/ideas/Set-up-portfolio"
        onClick={[Function]}
      >
         Details 
      </a>
    </li>
    <li
      className="ideaListing"
    >
      <h3>
        Summary: 
        Rebuild Media Logger
         
      </h3>
      <p>
        <b>
          Description
        </b>
        : 
        Rebuild media library cataloger, using .NET Core and proper separation of concerns.
      </p>
      <p>
        <b>
          Status
        </b>
        : 
        To Do
      </p>
      <a
        href="/ideas/rebuild-media-logger"
        onClick={[Function]}
      >
         Details 
      </a>
    </li>
    <li
      className="ideaListing"
    >
      <h3>
        Summary: 
        Voting App/Library
         
      </h3>
      <p>
        <b>
          Description
        </b>
        : 
        Build an app to support polls using various methods of voting, including but not limited to IRV, score voting, and STV.
      </p>
      <p>
        <b>
          Status
        </b>
        : 
        To Do
      </p>
      <a
        href="/ideas/voting-app-library"
        onClick={[Function]}
      >
         Details 
      </a>
    </li>
    <li
      className="ideaListing"
    >
      <h3>
        Summary: 
        Recipe Database
         
      </h3>
      <p>
        <b>
          Description
        </b>
        : 
        Aside from just doing "allourrecipes.com", it'd also be nice if I can get epub export for the recipes. A new digital version of the Bowhay cookbook.
      </p>
      <p>
        <b>
          Status
        </b>
        : 
        To Do
      </p>
      <a
        href="/ideas/Set-up-recipe-database"
        onClick={[Function]}
      >
         Details 
      </a>
    </li>
  </ul>,
]
`;
